src/crypto/tls/ech.go

```diff
@@ -0,0 +1,902 @@
+		// fresh random value.
+		helloInner = new(clientHelloMsg)
+		*helloInner = *helloBase
+
+		// Set "encrypted_client_hello" with an empty payload.
+		helloInner.encryptedClientHelloOffered = true
+
+		// Ensure that only TLS 1.3 and above are offered.
+		if v := helloInner.supportedVersions; v[len(v)-1] < VersionTLS13 {
+			return nil, nil, errors.New("tls: ech: TLS 1.3 required")
+		}
+
+		// Set "random".
+		if len(c.ech.hrrInnerRandom) == 0 {
+			// Generate a fresh "random".
+			helloInner.random = make([]byte, 32)
+			_, err := io.ReadFull(config.rand(), helloInner.random)
+			if err != nil {
+				return nil, nil, errors.New("tls: short read from Rand: " + err.Error())
+			}
+		} else {
+			// After HRR, use the "random" sent in the first ClientHelloInner.
+			helloInner.random = c.ech.hrrInnerRandom
+		}
+
+		// Construct the EncodedClientHelloInner.
+		//
+		// NOTE(cjpatton): It would be nice to incorporate more extensions, but
+		// "key_share" is the last extension to appear in the ClientHello before
+		// "pre_shared_key". As a result, the only contiguous sequence of outer
+		// extensions that contains "key_share" is "key_share" itself. Note that
+		// we cannot change the order of extensions in the ClientHello, as the
+		// unit tests expect "key_share" to be second to last extension.
+		outerExtensions := []uint16{extensionKeyShare}
+		if testingECHOuterExtMany {
+			// NOTE(cjpatton): Incorporating this particular sequence does not
+			// yield significant savings. However, it's useful to test that our
+			// server correctly handles a sequence of compressed extensions and
+			// not just one.
+			outerExtensions = []uint16{
+				extensionStatusRequest,
+				extensionSupportedCurves,
+				extensionSupportedPoints,
+			}
+		} else if testingECHOuterExtNone {
+			outerExtensions = nil
+		}
+		encodedHelloInner, ok := echEncodeClientHelloInner(helloInner.marshal(), outerExtensions)
+		if !ok {
+			return nil, nil, errors.New("tls: ech: encoding of EncodedClientHelloInner failed")
+		}
+
+		// Construct the ClientHelloOuter.
+		//
+		// Generate a fresh ClientHello, but replace "key_share" and "random",
+		// and "session_id" with the values generated for helloBase and set
+		// "server_name" to be the client-facing server.
+		hello, _, err = c.makeClientHello(config.MinVersion)
+		if err != nil {
+			return nil, nil, fmt.Errorf("tls: ech: %s", err)
+		}
+
+		// Set "random".
+		hello.random = helloBase.random
+
+		// Set "session_id" to be the same as ClientHelloInner.
+		hello.sessionId = helloBase.sessionId
+
+		// Set "key_share" to the same as ClientHelloInner.
+		hello.keyShares = helloBase.keyShares
+
+		// Set "server_name" to be the client-facing server.
+		hello.serverName = hostnameInSNI(string(echConfig.rawPublicName))
+
+		// Encrypt EncodedClientHelloInner.
+		//
+		// AEAD encryption authenticates the ClientHelloOuter sans the
+		// "encrypted_client_hello" extension.
+		helloOuterAad := hello.marshal()
```

Comment:
What do you mean by "determinsitic"? Does `marshal` flip coins? No :) But you have to be careful when calling `hello.marshal()` after mutating `hello`. To save some computation, `hello.marshal()` first checks if `hello.raw == nil`. If so, then it computes the serialized ClientHello and sets `hello.raw` to it. Otherwise it returns `hello.raw`.

Reply:
Right -- no coin flips. That's important as we call it multiple times and we need it to produce the same output.